### PR TITLE
feat(Android): Added longPressAndDrag action

### PR DESCRIPTION
**Pull request by gosha212**
## Description

- This pull request addresses the issue described here: #4003

In this pull request, I have added support for longPressAndDrag on Android.
Change the drag and drop test screen

---

> _For features/enhancements:_
 - [ ] I have added/updated the relevant references in the [documentation](https://github.com/wix/Detox/tree/master/docs) files.

> _For API changes:_
 - [ ] I have made the necessary changes in the [types index](https://github.com/wix/Detox/blob/master/detox/index.d.ts) file.

### CODE REVIEW

**Review by noomorph — APPROVED** [review]
Funny testing screen. I dig it.

**Review by noomorph — CHANGES_REQUESTED** [review]
@gosha212 please replace the graphical assets with other ones licensed under Creative Commons.

File: detox/detox.d.ts [hunk 1]
```diff
@@ -1740,9 +1740,11 @@ declare global {
             value?: unknown;
         }
 
-        interface IosElementAttributeFrame {
-            y: number;
+        interface IosElementAttributeFrame extends ElementAttributeFrame { }
+
+        interface ElementAttributeFrame {
             x: number;
+            y: number;
```
Comment (d4vidi) [inline detox/detox.d.ts:1747]: This is okay but can be better. I think we need to either get rid of `IosElementAttributeFrame` (and use `ElementAttributeFrame`, strictly) or introduce `AndroidElementAttributeFrame`. The former might be a breaking change, though. @noomorph wdyt?

File: detox/android/detox/src/main/java/com/wix/detox/espresso/scroll/LinearSwiper.kt [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+/**
+ * The delay between each motion event.
+ * Reducing this value may fail the swipe on different devices. Please change with caution.
+ */
+private const val EVENT_DELAY = 20L
+
+/**
+  * Implementation of @see DetoxSwiper that swipes in a linear fashion uses const delay between events.
+ */
+class LinearSwiper @JvmOverloads constructor(
```
Comment (d4vidi) [inline detox/android/detox/src/main/java/com/wix/detox/espresso/scroll/LinearSwiper.kt:16]: That's great and all (it really is!), but why the dup code? It's identical to `FlinglessSwiper` in practically everything but `calcEventTime()`, isn't it? I think there's room for sharing everything we can via a base class 🤔

File: detox/android/detox/src/main/java/com/wix/detox/espresso/action/DetoxSwipeWithLongPress.kt [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+                    if (!moveTo(targetX, targetY)) {
+                        return
+                    }
+                }
+            } finally {
+                moveTo(endX, endY)
+                wait(durationEnd)
+                finishAt(endX, endY)
+            }
+        }
```
Comment (d4vidi) [inline detox/android/detox/src/main/java/com/wix/detox/espresso/action/DetoxSwipeWithLongPress.kt:42]: Pretty elegant!

File: detox/android/detox/src/full/java/com/wix/detox/espresso/DetoxAction.java [hunk 4]
```diff
@@ -182,6 +184,31 @@ public static ViewAction adjustSliderToPosition(final Float newPosition) {
...
+                                              Double normalizedPositionY,
+                                              ViewInteraction targetElement,
+                                              Double normalizedTargetPositionX,
+                                              Double normalizedTargetPositionY,
+                                              boolean isFast,
+                                              Integer holdDuration) {
+
+        Log.d("DetoxAction", "Current thread is: " + Thread.currentThread().getName());
+
+        View targetView = ViewInteractionExtKt.getView(targetElement);
```
Comment (d4vidi) [inline detox/android/detox/src/full/java/com/wix/detox/espresso/DetoxAction.java:198]: Could you add a comment explaining this (understandable) workaround?

File: detox/ios/Detox/AppleTouchEvents/DTXTouchInjector.m [hunk 5]
```diff
@@ -24,6 +24,9 @@
 #import "DTXRunLoopSpinner.h"
 #import "DTXTouchInfo-Private.h"
 
+DTX_CREATE_LOG(DTXTouchInjector)
```
Comment (d4vidi) [inline detox/ios/Detox/AppleTouchEvents/DTXTouchInjector.m:27]: 🤔

File: detox/test/e2e/28.drag-and-drop.test.js [hunk 6]
```diff
@@ -1,35 +1,77 @@
...
+
+describe('Drag And Drop', () => {
   beforeEach(async () => {
     await device.reloadReactNative();
     await element(by.text('Drag And Drop')).tap();
   });
 
   afterEach(async () => {
-    await element(by.id('closeButton')).tap();
   });
```
Comment (d4vidi) [inline detox/test/e2e/28.drag-and-drop.test.js:10]: So I guess we can ditch the `afterEach`, then?

**Review by d4vidi — CHANGES_REQUESTED** [review]
Hey solid work! 👏🏻 Requesting subtle code changes (see comments) before approval!

**Comment by asafkorem** [comment]
@gosha212 we generally prefer semantic commit message for the PR names, see:
https://gist.github.com/joshbuchea/6f47e86d2510bce28f8e7f42ae84c716

for example: "feat(Android): implement `longPressAndDrag()` action."

File: detox/ios/Detox/Actions/NSObject+DetoxActions.m [hunk 7]
```diff
@@ -122,6 +124,12 @@ - (void)dtx_longPressAtPoint:(CGPoint)normalizedPoint duration:(NSTimeInterval)d
 	
 	// Add end point
 	[points addObject:@(endPoint)];
+	// This is an ugly hack needed to solve a mystery with a missing last point in the trajectory. We add one more point with the same delta after the end. The OS will happily ignore it.
+	// It is criticul to add a point with the same pace in order to insure it works for all cases. If you use point with a different direction or just (0,0) the OS will not ignore it.
+	CGPoint point = CGPointMake(startPoint.x + (numOfPoints + 1) * xDiffDelta, startPoint.y + (numOfPoints + 1) * yDiffDelta);
+	[points addObject:@(point)];
```
Comment (asafkorem) [inline detox/ios/Detox/Actions/NSObject+DetoxActions.m:130]: hmm, let's test this with different iOS versions (to make sure it works for 15, 16, 17). Can you verify that?

File: detox/ios/Detox/Actions/NSObject+DetoxActions.m [hunk 8]
```diff
@@ -122,6 +124,12 @@ - (void)dtx_longPressAtPoint:(CGPoint)normalizedPoint duration:(NSTimeInterval)d
...
+		
+	dtx_log_info(@"longPressAtPoint Normalized start point %@, Normalized end point %@", NSStringFromCGPoint(calcNormalizedPoint), NSStringFromCGPoint(calcNormalizedTargetPoint));
```
Comment (asafkorem) [inline detox/ios/Detox/Actions/NSObject+DetoxActions.m:132]: no need

File: detox/ios/Detox/Actions/NSObject+DetoxActions.m [hunk 9]
```diff
@@ -19,6 +19,8 @@
 #import "UIView+DetoxUtils.h"
 #import "UIImage+DetoxUtils.h"
 
+DTX_CREATE_LOG(DetoxActions)
+
```
Comment (asafkorem) [inline detox/ios/Detox/Actions/NSObject+DetoxActions.m:23]: same

**Review by asafkorem — CHANGES_REQUESTED** [review]

File: detox/detox.d.ts (hunk 1 above)
Comment (noomorph) [reply detox/detox.d.ts:1747]: Indeed, we don't break typings to adhere to the semantic versioning policy. Do your best not to change the signatures dramatically. It's better to accumulate a list of breaking changes and do it in one focused leap.

File: detox/android/detox/src/main/java/com/wix/detox/espresso/scroll/LinearSwiper.kt (hunk 2 above)
Comment (gosha212) [reply detox/android/detox/src/main/java/com/wix/detox/espresso/scroll/LinearSwiper.kt:16]: very good point

File: detox/ios/Detox/Actions/NSObject+DetoxActions.m [hunk 10]
```diff
@@ -19,8 +19,7 @@
 #import "UIView+DetoxUtils.h"
 #import "UIImage+DetoxUtils.h"
 
-DTX_CREATE_LOG(DetoxActions)
-
+@implementation NSObject (DetoxActions)
 @implementation NSObject (DetoxActions)
```
Comment (asafkorem) [inline detox/ios/Detox/Actions/NSObject+DetoxActions.m:23]: you accidently duplicated this line here

**Review by asafkorem — CHANGES_REQUESTED** [review]

File: detox/ios/Detox/Actions/NSObject+DetoxActions.m [hunk 11]
```diff
@@ -19,6 +19,7 @@
 #import "UIView+DetoxUtils.h"
 #import "UIImage+DetoxUtils.h"
 
+@implementation NSObject (DetoxActions)
 @implementation NSObject (DetoxActions)
```
Comment (asafkorem) [inline detox/ios/Detox/Actions/NSObject+DetoxActions.m:23]:
```suggestion
@implementation NSObject (DetoxActions)
```

File: detox/detox.d.ts (hunk 1 above)
Comment (gosha212) [reply detox/detox.d.ts:1747]: Leaving it like that

**Review by d4vidi — APPROVED** [review]

**Pull request merged**